File: app/src/main/java/vergecurrency/vergewallet/service/model/wallet/BNTransaction.java
package vergecurrency.vergewallet.service.model.wallet;

import com.google.crypto.tink.subtle.Hex;

import io.horizontalsystems.bitcoinkit.io.BitcoinInput;
import io.horizontalsystems.bitcoinkit.managers.UnspentOutputProvider;
import io.horizontalsystems.bitcoinkit.managers.UnspentOutputSelector;
import io.horizontalsystems.bitcoinkit.models.TransactionInput;
import io.horizontalsystems.bitcoinkit.models.TransactionOutput;
import vergecurrency.vergewallet.helpers.utils.DataUtils;


public class BNTransaction {

	private String chain;
	private String network;
	private boolean coinbase;
	private int mintIndex;
	private String spentTxid;
	private String mintTxid;
	private long mintHeight;
	private long spentHeight;
	private String address;
	private String script;
	private long value;
	private long confirmations;



	//This can possibly be a big chunk of shit code. to test first priority.
	public UnspentTransaction asUnspentTransaction() throws InvalidScriptPubKeyHexException, InvalidTxIdHexException {
		byte[] lockingScript, txid;
		try {
			lockingScript = Hex.decode(script);
		} catch (Exception e) {
			throw new InvalidScriptPubKeyHexException(script);
		}
		try {
			 txid = Hex.decode(mintTxid);
		} catch (Exception e) {
			throw new InvalidTxIdHexException(mintTxid);
		}

		TransactionOutput transactionOutput = new TransactionOutput();
		transactionOutput.setValue(value);
		transactionOutput.setLockingScript(lockingScript);
		byte[] txHash = DataUtils.reverse(txid);
		TransactionOutPoint transactionOutPoint = new TransactionOutPoint(txHash,mintIndex);

		return new UnspentTransaction(transactionOutput,transactionOutPoint);
	}

	public TransactionInput asInputTransaction() throws InvalidTxIdHexException {
		byte[] txid;
		try {
			txid = Hex.decode(mintTxid);
		} catch (Exception e) {
			throw new InvalidTxIdHexException(mintTxid);
		}
		byte[] txHash = DataUtils.reverse(txid);
		TransactionOutPoint transactionOutPoint = new TransactionOutPoint(txHash,mintIndex);

		//This is probs bullshit but the BitcoinKit on Android and Swift don't correspond on this one so let's see...
		return new TransactionInput(new BitcoinInput(transactionOutPoint.serialize()));

	}


	class InvalidScriptPubKeyHexException extends Exception {
		public InvalidScriptPubKeyHexException (String hex) {
			super(hex);
		}
	}

	class InvalidTxIdHexException extends Exception {
		public InvalidTxIdHexException (String hex) {
			super(hex);

		}
	}

}

File: app/src/main/kotlin/vergecurrency/vergewallet/service/model/wallet/BNTransaction.kt
package vergecurrency.vergewallet.service.model.wallet

import com.google.gson.Gson
import io.horizontalsystems.bitcoinkit.io.BitcoinInput
import io.horizontalsystems.bitcoinkit.models.TransactionInput
import io.horizontalsystems.bitcoinkit.models.TransactionOutPoint
import io.horizontalsystems.bitcoinkit.models.TransactionOutput
import org.bouncycastle.util.encoders.Hex
import vergecurrency.vergewallet.helpers.utils.DataUtils


class BNTransaction {

    private val chain: String? = null
    private val network: String? = null
    private val coinbase: Boolean = false
    private val mintIndex: Int = 0
    private val spentTxid: String? = null
    private val mintTxid: String? = null
    private val mintHeight: Long = 0
    private val spentHeight: Long = 0
    private val address: String? = null
    private val script: String? = null
    private val value: Long = 0
    private val confirmations: Long = 0

    companion object {

        fun decode(JSON: String): BNTransaction {
            return Gson().fromJson(JSON, BNTransaction::class.java)
        }
    }


    //This can possibly be a big chunk of shit code. to test first priority.
    @Throws(InvalidScriptPubKeyHexException::class, InvalidTxIdHexException::class)
    fun asUnspentTransaction(): UnspentTransaction {
        val lockingScript: ByteArray
        val txid: ByteArray
        try {
            lockingScript = Hex.decode(script!!)
        } catch (e: Exception) {
            throw InvalidScriptPubKeyHexException(script!!)
        }

        try {
            txid = Hex.decode(mintTxid!!)
        } catch (e: Exception) {
            throw InvalidTxIdHexException(mintTxid!!)
        }

        val transactionOutput = TransactionOutput()
        transactionOutput.value = value
        transactionOutput.lockingScript = lockingScript
        val txHash = DataUtils.reverse(txid)
        val transactionOutPoint = TransactionOutPoint(txHash, mintIndex)

        return UnspentTransaction(transactionOutput, transactionOutPoint)
    }

    @Throws(InvalidTxIdHexException::class)
    fun asInputTransaction(): TransactionInput {
        val txid: ByteArray
        try {
            txid = Hex.decode(mintTxid!!)
        } catch (e: Exception) {
            throw InvalidTxIdHexException(mintTxid!!)
        }

        val txHash = DataUtils.reverse(txid)
        val transactionOutPoint = TransactionOutPoint(txHash, mintIndex)

        //This is probs bullshit but the BitcoinKit on Android and Swift don't correspond on this one so let's see...
        return TransactionInput(BitcoinInput(transactionOutPoint.serialize()))

    }


    internal inner class InvalidScriptPubKeyHexException(hex: String) : Exception(hex)

    internal inner class InvalidTxIdHexException(hex: String) : Exception(hex)

}
